### PR TITLE
IGNITE-17014 [Native Persistence 3.0] Porting FilePageStoreFactory and FilePageStore from 2.0

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-17014

### CODE REVIEW

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java [hunk 1]
```diff
@@ -198,7 +198,7 @@ public static int getVersion(long pageAddr) {
      * @param pageAddr Page address.
      * @param ver Version.
      */
-    protected static void setVersion(long pageAddr, int ver) {
+    public static void setVersion(long pageAddr, int ver) {
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java:201]: I'm sorry, why does it need to be public?

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+            assert pageBuf.position() == 0;
+            assert pageBuf.order() == nativeOrder();
+            assert off <= allocatedBytes.get() : "calculatedOffset=" + off
+                    + ", allocated=" + allocatedBytes.get() + ", headerSize=" + headerSize() + ", filePath=" + filePath;
+
+            int n = readWithFailover(pageBuf, off);
+
+            // If page was not written yet, nothing to read.
+            if (n < 0) {
+                pageBuf.put(new byte[pageBuf.remaining()]);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java:204]: This is interesting, how can that happen? And why do we have to write zeroes

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+        ByteBuffer headerBuffer = ByteBuffer.allocate(headerSize()).order(nativeOrder());
+
+        fileIo.readFully(headerBuffer);
+
+        checkHeader(headerBuffer);
+
+        long fileSize = filePath.toFile().length();
+
+        // Every file has a special meta page.
+        if (fileSize == headerSize()) {
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java:605]: This whole "if" can just be replaces with "fileSize -= headerSize();"

Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java:605]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java (hunk 2 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java:204]: -1 if the position is greater than or equal to the file's current size.
Removed nulling.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java (hunk 1 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/io/PageIo.java:201]: For **FilePageStoreTest#createPageByteBuffer**

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java (hunk 2 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStore.java:204]: In personal correspondence, we decided to leave it as it is, added some documentation.

**Comment by tkalkirill** [comment]
Please make code review.

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStoreTest.java [hunk 4]
```diff
@@ -498,8 +498,7 @@ private static ByteBuffer copyOf(ByteBuffer buffer) {
     private static ByteBuffer createPageByteBuffer() {
         ByteBuffer buffer = ByteBuffer.allocateDirect(PAGE_SIZE).order(nativeOrder());
 
-        PageIo.setType(bufferAddress(buffer), MAX_IO_TYPE);
-        PageIo.setVersion(bufferAddress(buffer), 1);
+        new TestPageIo().initNewPage(GridUnsafe.bufferAddress(buffer), 0, PAGE_SIZE);
```
Comment (ibessonov) [inline modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStoreTest.java:501]: I bet we already have a test io somewhere nearby, can we reuse it? Or they're all private?

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 5]
```diff
@@ -265,13 +266,29 @@ public static String toHexString(long addr, int len) {
...
 
+    /**
+     * Returns hex representation of memory region.
+     *
+     * @param buf Buffer which content should be converted to string.
+     */
+    public static String toHexString(ByteBuffer buf) {
+        StringBuilder sb = new StringBuilder(buf.capacity() * 2);
+
+        for (int i = 0; i < buf.capacity(); i++) {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:284]: Should't you start with position and end with limit? Please reflect in javadoc that position and limit are ignored and the whole data chunk is converted to string

File: modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStoreTest.java (hunk 4 above)
Comment (tkalkirill) [reply modules/page-memory/src/test/java/org/apache/ignite/internal/pagememory/persistence/store/FilePageStoreTest.java:501]: Reuse **org.apache.ignite.internal.pagememory.TestPageIoModule.TestPageIo**

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java (hunk 5 above)
Comment (tkalkirill) [reply modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:284]: You're right, corrected and expanded the tests.

**Comment by tkalkirill** [comment]
@ibessonov Please make code review.

**Pull request merged**